stellargraph/data/unsupervisedSampler.py

```diff
@@ -0,0 +1,229 @@
+            self.walker = UniformRandomWalk(G)
+
+        if nodes is None:
+            self.nodes = (
+                G.nodes()
+            )  # if no root nodes are provided for sampling defaulting to using all nodes as root nodes.
+        elif not is_real_iterable(nodes):  # check whether the nodes provided are valid.
+            raise ValueError("nodes parameter should be an iterableof node IDs.")
+        else:
+            self.nodes = nodes
+
+        if (
+            length < 2
+        ):  # for a  sample pair need at least one neighbor beyond the root node.
+            raise ValueError(
+                "({}) For generating (target,context) samples, walk length has to be at least 2".format(
+                    type(self).__name__
+                )
+            )
+        else:
+            self.length = length
+
+        if number_of_walks < 1:
+            raise ValueError(
+                "({}) At least 1 walk from each head node has to be done".format(
+                    type(self).__name__
+                )
+            )
+        else:
+            self.number_of_walks = number_of_walks
+
+        self.seed = seed
+
+    def generator(self, batch_size):
+
+        """
+        This method yields a batch_size of +/- samples when it is called. The batch_size must be provided. It should be a positive integer. Moreover, for now we make the batch_size even by
+        unit incrementing it, incase it is odd. This is because we generate one negative sample for each positive sample.
+        This method:
+            1. Iterates over the root nodes.
+                1. Generates one walk at a time based on the walker method of the class  over the given Stellargraph. 
+                2. Incrementally extract positive (target,context) from the walks.
+                    1. For each positive pair sample a negative (target, context) from the graph based on a sampling distribution.
+                3. Once the total positive and negative samples generated are equal to the batch_size, the generator returns the samples. 
+         The positive samples are identical to GraphSAGE way of sampling. The negative sampling uses identical sampling distribution. 
+         However, no. of negative samples generated is perhaps different. This observation is based on our present understanding of Hamilton's implementation of GraphSAGE from github.
+         Note, we do not check if the negative sample is a positive context or a target itself. Although we do check that target is not its own context.
+         
+         Args:
+             batch_size: <int> number of samples to generate in each call of generator
+
+        Returns:
+            A tuple of lists: (target, context) pairs and their labels. 
+        """
+
+        self._check_parameter_values(batch_size=batch_size)
+
+        # making the batch_size even since we generate 1 negative sample for each positive one.
+        if batch_size % 2 != 0:
+            batch_size += 1
```

Comment:
I think I'd prefer raising an error here, so that people aren't confused by their batch sizes being different than expected.

Reply:
Fixed this. Now, the generator works with only even sized batches. Added unit test as well.